### ISSUE TITLE
`git://` protocol support

### ISSUE THREAD

**Issue by 5paceToast** [isomorphic-git/isomorphic-git#665]
I have a project that uses one of your downstreams (antora).
Due to current infrastructure limitations (gitolite + cgit), there are three ways to fetch a project:

1. Using an ssh key (no usernames / passwords / oauth) - obviously not ok, since mirroring is meant to be allowed.
2. "dumb" http (cgit) - something that is incompatible with current isomorphic-git.
3. the `git://` protocol - which does not seem to be supported right now

The error when attempting 3 is `error: Content source uses an unsupported transport protocol: git://[...]`.

Please add support to the `git` protocol, as it makes isomorphic-git impossible to use under the circumstances listed above, as well as similar ones (e.g ssh+git-daemon only setups, etc).

**Comment by 5paceToast** [comment]
> Here's a test repository to use (or any other git repo for Alpine) for testing cgit over https.
> https://git.alpinelinux.org/cgit/docs/docs.a.o/

This gives out an error stating `Error: Remote does not support the "smart" HTTP protocol, and isomorphic-git does not support the "dumb" HTTP protocol, so they are incompatible: https://[...]`.

As I mentioned here:
> "dumb" http (cgit) - something that is incompatible with current isomorphic-git.

Our cgit only knows "dumb" http right now.
The infrastructure team is looking into adding "smart" http support, but there are likely many setups where it is unavailable.

> Is this really needed? Would it be better to focus on dumb http support? (whatever dumb http is)

That is certainly possible, but currently thw two are marked as explicitly incompatible, while `git://` is simply missing.
Further, in many corporate setups, `git://` is used in preference of `http` (for whatever reason).

**Comment by mojavelinux** [comment]
I now understand the dumb http means a static web server. Hence, no cgi-like capabilities to prepare a pack file.

See https://git-scm.com/book/en/v2/Git-on-the-Server-Smart-HTTP for details.

> Further, in many corporate setups, git:// is used in preference of http (for whatever reason).

This has not been my experience. Not saying you're wrong or anything, just pointing out that it's not universal.

**Comment by wmhilton** [comment]
The `git://` has zero password protection... I guess if it's behind a corporate VPN then it wouldn't matter. So the long answer is supporting `git://` is probably possible... but because it relies on TCP port 9418, there's no way to make it work in the browser. Therefore it would be a Node-only feature and probably supported via a plugin, e.g.
```js
// pseudo-code
const tcpgit = require('...')
git.plugins.get('protocols').set('git:', tcpgit)
```

Can you clarify some things about the infrastructure @5paceToast so I know we're talking about the same things? 
- cgit refers to [this project](https://git.zx2c4.com/cgit/about/)?
- gitolite refers to [this project](http://gitolite.com/gitolite/index.html)?

I see that `gitolite` supports the smart HTTP protocol but requires some [scary configuration](http://gitolite.com/gitolite/http.html). The [setup script](https://github.com/sitaramc/gitolite/blob/master/t/smart-http.root-setup) presumes an Apache webserver. Do you know whether your infrastructure is using Apache?

(And yes, I'm obviously trying to think of ways to resolve the situation without me having to write more code 😆 But if we can't figure out how to enable Smart HTTP on your server I guess we'll have to implement `git://`)

**Comment by 5paceToast** [comment]
> The `git://` has zero password protection...

The git protocol is *primarily* intended to be read-only. It is possible to make it `r/w`, as I've seen done behind company firewalls, and `git(1)` does support using it as `r/w`, but it's almost always `ro`.
> So the long answer is supporting git:// is probably possible... but because it relies on TCP port 9418, there's no way to make it work in the browser. Therefore it would be a Node-only feature and probably supported via a plugin

My use-case is with antora - a nodejs-based documentation-centric static site generator. I can see the issue though.
> - cgit refers to this project?
> - gitolite refers to this project?

Both correct - my understanding is that gitolite is only used for ssh clients, http is provided by cgit.
> I see that gitolite supports the smart HTTP protocol but requires some scary configuration.

I have doubts regarding this being implemented, it would at a minimum require some larger reconfiguration (currently, port 80 is provided entirely by cgit, as it provides "dumb" http and a web interface).
> Do you know whether your infrastructure is using Apache?

The infrastructure is behind a reverse-proxy (which nmap reveals to likely be nginx).
I don't know exactly what's behind the reverse-proxy in this specific case, but I was told that for "simple" httpds we use darkhttpd.
Since cgit is a CGI-based application, however, it is more likely we use nginx behind the reverse proxy as well.
> But if we can't figure out how to enable Smart HTTP on your server

It's currently being looked into, but there's also a (very) late release incoming, so there's no real ETA on that either - manhour count is a real problem.

Obviously, implementing "dumb" http (as per above conversation) is fine as well, and (in theory) both should be implemented at some point (at the very least for `isogit(1)` -> `git(1)` compatibility (you make a good point re: browser usage)).

**Comment by wmhilton** [comment]
Hmm. The reason I haven't really considered implementing the "dumb" HTTP protocol is it's really inefficient. Github [disabled](https://blog.github.com/2011-03-09-git-dumb-http-transport-to-be-turned-off-in-90-days/) cloning via dumb http back in 2011, mentioning that less than 7% of traffic used it. So I figured nowadays, the usage must be near 0%.

Could you do a quick test and see whether cloning your particular repo via dumb `http://` is waaaay slow compared to cloning over `git://`? Because the nice thing about dumb HTTP would be it is isomorphic (e.g. since it's still HTTP we can still do it in the browser!). But if it's super slow then it probably makes more sense to do the extra work to expose the remote protocol handlers as plugins. My guess is that `git://` is going to be so much faster and more reliable than dumb HTTP that it's a clear winner... but I'm down to implement either. (I'd implement both if I had infinite time! But it makes sense to test and see which is better, then prioritize.)

**Comment by 5paceToast** [comment]
For github, I highly suspect it's 0%, but there *are* other git hosts out there besides github/gitlab :)
For my repo, it doesn't make a significant difference (based on below calculations).
Here's a (relatively short) analysis of git clone performance:

## Abstract

It is suspected that due to its lack of ability to prepare custom packs, the "dumb http" git protocol will perform worse than `git-daemon`, as well as be less reliable.
We test the former and discuss the latter.
The data indicates that the flat performance hit due to using git is greater than the one for dumb http (though both are negligible), but that git scales significantly better for larger repositories.

## Methodology

We take two repositories: Alpine's [user-handbook](https://git.alpinelinux.org/cgit/docs/user-handbook/) repository, and the [linux kernel](https://github.com/torvalds/linux/).

We then create a directory into which we locally cache both of them:
```sh
mkdir ~/git
cd ~/git
git clone --bare git://git.alpinelinux.org/docs/user-handbook.git
git clone --bare https://github.com/torvalds/linux.git
```

We then enable "dumb http" support:
```sh
for f in user-handbook linux
do
  cd $f.git
  mv hooks/post-update.sample hooks/post-update
  chmod +x hooks/post-update
  ./hooks/post-update
  cd ..
done
```

Then we run a static http server and git-daemon (note: the following two commands are ran in separate terminals):
```sh
python -m http.server
git daemon --base-path=. --export-all --reuseaddr --informative-errors
```

Each repository will be cloned 3 times, into `tmpfs`, each of which will be timed, using the following commands:
```sh
c_git() { time git clone -q git://localhost/$1 $1-git && rm -rf $1-git }
c_http() { time git clone -q http://localhost:8000/$1 $1-http && rm -rf $1-http }
```

This approach means that we can eliminate variables such as read/write speed (reads are from cache, writes are to ram), network speed (everything happens over lo) and similar - allowing us to measure specifically protocol overhead.

## Data

### Small Repository with Git
|Run Number|Time (s)|% Relative to Mean|
|-|-:|:-|
|1|0.072|104%|
|2|0.063|91%|
|3|0.073|106%|
|Mean|0.069|100%|
|Sigma|0.0043|6.2%|

### Small Repository with Dumb HTTP
|Run Number|Time (s)|% Relative to Mean|
|-|-:|:-|
|1|0.032|94%|
|2|0.032|94%|
|3|0.038|112%|
|Mean|0.034|100%|
|Sigma|0.0026|7.6%|

### Large Repository with Git
|Run Number|Time (s)|% Relative to Mean|
|-|-:|:-|
|1|363.03|99.9%|
|2|366.22|100.8%|
|3|360.52|99.2%|
|Mean|363.26|100%|
|Sigma|1.98|0.5%|

### Large Repository with Dumb HTTP
|Run Number|Time (s)|% Relative to Mean|
|-|-:|:-|
|1|459.26|100.9%|
|2|452.58|99.5%|
|3|452.48|99.5%|
|Mean|454.77|100%|
|Sigma|2.99|0.7%|

### Mean Summary and Comparisons
|Checkout Size|Checkout Type|Time (s)|% Relative to Alternative Type|
|-|-|-:|:-|
|Small|Git|0.069|203%|
|Small|HTTP|0.034|49%|
|Large|Git|363.26|80%|
|Large|HTTP|454.77|125%|

## Analysis and Conclusion

The data shows that "git" has a significant initial overhead cost, but that it scales significantly better than "dumb http".
However, the scaling, while linear, appears to be lower than 1:1 - meaning that as repositories get larger and larger, this becomes less important, though this may be due to io bottlenecking (even on tmpfs).
It is also notable that git has reliably smaller standard deviations, suggesting it is more consistent.

It is doubtful that repositories will get much larger than 2gb, so we can consider http overhead over git to be at least 25%,
While git is shown as slower for smaller repositories, that difference is mostly negligible.
As such, the `git-daemon`-based protocol should be preferred over "dumb" http for read operations.

## Additional Notes Regarding Reliability
Observing the behavior of the HTTP server, we can see that each object is downloaded separately using HTTP GET.
This would normally not be a problem, but because of the nature of large repositories, it downloads the pack like this - something that will not deal well with packet loss.
Whether or not this applies to the git protocol is unknown, and should be investigated separately.

## Issues
Unfortunately, it is not possible to calculate initial per-protocol overhead, nor graph the increase in time based on commit-byte.
This is because git does not offer a protocol-less cloning mechanism (my understanding is that the file-based one is still greater than cp).
If one wanted to make this more rigorous, one would write an extension to git-clone that would only perform cp(1), and use that as the control, as well as making a single-commit single-empty-file repository to get a reliable 0.
With that, it would become possible to calculate and plot the actual protocol overhead / commit-byte.
There's also a lack of sample data - this should be repeated with a statistically significant, randomly selected set of repositories (but I'm lazy and time constrained).

## Implementation Notes
All of the above tests were ran on a i5-5200U laptop with 8GB of ram, that was otherwise idle.
If attempting to reproduce, I recommend adjusting system fd limit, and increasing filetree caching aggressiveness - as well as minimizing swappiness, to avoid hitting unnecessary IO.
If one happens to have additional RAM, everything should be in tmpfs, and the `rm -rf` step can be skipped (don't forget to increment indexes in that case).

**Comment by mojavelinux** [comment]
@5paceToast Tremendous analysis. Thanks for sharing. I'm so impressed and I learned a lot.

> The data shows that "git" has a significant initial overhead cost, but that it scales significantly better than "dumb http".

Fascinating. And it kind of makes sense that using dumb http would incur less overhead for smaller throughputs.

It seems like the ability to clone over dumb HTTP is worth having since it holds up compared to smart HTTP and increases compatibility. I could also see it useful in applications that serve data back and forth using git over a simple Node HTTP server. So it might even have application value beyond just compatibility.

**Comment by wmhilton** [comment]
@5paceToast Awesome analysis!

I didn't realize that the "dumb" protocol still used packfiles. That's really interesting, because there's no way of knowing which packfiles contain which commits unless you download the .idx files... which I guess you could do and download the packfiles themselves as needed? Or (more reliable) just download all the packfiles listed in /objects/info/packs.

I will point out one significant flaw in the methodology:

> This approach means that we can eliminate variables such as ... network speed (everything happens over lo)

The dumb protocol will have to create a new HTTP request for every file - which involves a new TCP connection - and if the server is behind HTTPS a TLS handshake for each connection. This involves several "round trips" between the client and the server. On a localhost, the distance is 0km so the time per round trip time (RTT) is near 0ms. But in a realistic scenario, the RTT could be tens of milliseconds. So in a localhost test, you might observe similar clone speeds between git:// with it's single TCP connection and dumb HTTP 1000 connections. But once you add 10ms of latency, the overhead of git:// is 10ms but for dumb HTTP could be 1000×10ms or 10 seconds. I was much more worried about this when I thought dumb HTTP downloaded every single commit, tree, and blob in loose format. If that were the case the overhead from HTTP requests would scale with the size of the repository. Now that I know it downloads packfiles, I imagine the number of requests is nowhere near 1000. The overhead from HTTP requests then scales at worst with the number of times git push is called (if a new packfile is created for each push) or even less than that if `git-update-server` is smart enough to garbage collect every now and then.

In conclusion... I think we should add support for cloning over dumb HTTP!

**Comment by 5paceToast** [comment]
> The dumb protocol will have to create a new HTTP request for every file - which involves a new TCP connection - and if the server is behind HTTPS a TLS handshake for each connection. This involves several "round trips" between the client and the server. On a localhost, the distance is 0km so the round trip time (RTT) is near 0ms. But in a realistic scenario, the RTT could be tens of milliseconds.

Well the goal was to measure the protocol specifically (i.e packfile support, etc) :)
And as we can see, even *with* the lack of roundtrip here, http still ends up lagging behind for progressively larger repositories.

I'll add a new issue for "dumb http" support and mention this issue.
This issue can stick around as a tracker bug for `git://` support (which will happen eventually, but likely (potentially much) later).

**Comment by eddo888** [comment]
i would also like git:// protocol support, it may require you to add ssh:// as one relies on the other and the ssh provides key/secret pubkey authentication

**Comment by mojavelinux** [comment]
I don't think the git protocol relies on SSH. What I do know is that the git URL is sometimes rewritten as an SSH URL, which then uses the SSH rather than the git protocol.

In the git clone docs it even says:

"a URL like "git://example.org/path/to/repo.git" will be rewritten to "ssh://example.org/path/to/repo.git" for pushes, but pulls will still use the original URL"

**Comment by 5paceToast** [comment]
`git://` has nothing in common with ssh.
The git protocol is implemented by `git-daemon(1)`, listens on a separate TCP port, and has the path rewritten based on the arguments provided to it (e.g `--base-path` and co, see the example benchmark above for how to invoke it).
The ssh protocol is implemented similar to local files (e.g `git clone /srv/something`), with the path being what goes after the : (in `user@host:path` format) or after the first non-protocol / (in `ssh://host/path` format).

The `git://` protocol provides no authentication whatsoever, and intentionally so.

For more details on `git://`, please see https://git-scm.com/docs/git-daemon.

**Comment by mojavelinux** [comment]
FYI, GitHub is disabling access to repositories via the git protocol for security reasons. Just something to note. https://github.blog/2021-09-01-improving-git-protocol-security-github/
